awss3/store.go

```diff
@@ -0,0 +1,703 @@
+	}
+
+	params := &s3.ListObjectsInput{
+		Bucket:    aws.String(f.bucket),
+		MaxKeys:   &itemLimit,
+		Prefix:    &q.Prefix,
+		Delimiter: &q.Delimiter,
+	}
+
+	folders := make([]string, 0)
+
+	for {
+		select {
+		case <-ctx.Done():
+			// If has been closed
+			return folders, ctx.Err()
+		default:
+			if q.Marker != "" {
+				params.Marker = &q.Marker
+			}
+			resp, err := f.client.ListObjectsWithContext(ctx, params)
+			if err != nil {
+				return nil, err
+			}
+			for _, cp := range resp.CommonPrefixes {
+				folders = append(folders, strings.Trim(*cp.Prefix, `/`))
+			}
+			return folders, nil
+		}
+	}
+}
+
+/*
+// Copy from src to destination
+func (f *FS) Copy(ctx context.Context, src, des cloudstorage.Object) error {
+
+	so, ok := src.(*object)
+	if !ok {
+		return fmt.Errorf("Copy source file expected s3 but got %T", src)
+	}
+	do, ok := des.(*object)
+	if !ok {
+		return fmt.Errorf("Copy destination expected s3 but got %T", des)
+	}
+
+	oh := so.b.Object(so.name)
+	dh := do.b.Object(do.name)
+
+	_, err := dh.CopierFrom(oh).Run(ctx)
+	return err
+}
+
+// Move which is a Copy & Delete
+func (f *FS) Move(ctx context.Context, src, des cloudstorage.Object) error {
+
+	so, ok := src.(*object)
+	if !ok {
+		return fmt.Errorf("Move source file expected s3 but got %T", src)
+	}
+	do, ok := des.(*object)
+	if !ok {
+		return fmt.Errorf("Move destination expected s3 but got %T", des)
+	}
+
+	oh := so.b.Object(so.name)
+	dh := do.b.Object(des.name)
+
+	if _, err := dh.CopierFrom(oh).Run(ctx); err != nil {
+		return err
+	}
+
+	return oh.Delete(ctx)
+}
+*/
+// NewReader create GCS file reader.
+func (f *FS) NewReader(o string) (io.ReadCloser, error) {
+	return f.NewReaderWithContext(context.Background(), o)
+}
+
+// NewReaderWithContext create new File reader with context.
+func (f *FS) NewReaderWithContext(ctx context.Context, objectname string) (io.ReadCloser, error) {
+	res, err := f.client.GetObjectWithContext(ctx, &s3.GetObjectInput{
+		Key:    aws.String(objectname),
+		Bucket: aws.String(f.bucket),
+	})
+	if err != nil {
+		// translate the string error to typed error
+		if strings.Contains(err.Error(), "NoSuchKey") {
+			return nil, cloudstorage.ErrObjectNotFound
+		}
+		return nil, err
+	}
+	return res.Body, nil
+}
+
+// NewWriter create Object Writer.
+func (f *FS) NewWriter(objectName string, metadata map[string]string) (io.WriteCloser, error) {
+	return f.NewWriterWithContext(context.Background(), objectName, metadata)
+}
+
+// NewWriterWithContext create writer with provided context and metadata.
+func (f *FS) NewWriterWithContext(ctx context.Context, objectName string, metadata map[string]string) (io.WriteCloser, error) {
+
+	// Create an uploader with the session and default options
+	uploader := s3manager.NewUploader(f.sess)
+
+	pr, pw := io.Pipe()
+
+	go func() {
+		// TODO:  this needs to be managed, ie shutdown signals, close, handler err etc.
+
+		// Upload the file to S3.
+		_, err := uploader.Upload(&s3manager.UploadInput{
```

Comment:
could we use the `UploadWithContext` method instead, so that we don't leak this go routine if the ctx is canceled? 
https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#Uploader.UploadWithContext

Reply:
good idea, done